autogen/agentchat/contrib/capabilities/context_handling.py

```diff
@@ -100,9 +109,11 @@ def truncate_str_to_tokens(text: str, max_tokens: int) -> str:
     Returns:
         Truncated string.
     """
-    truncated_string = ""
-    for char in text:
-        truncated_string += char
-        if token_count_utils.count_token(truncated_string) == max_tokens:
-            break
-    return truncated_string
+
+    tokens = text.split()
+    for token_count in range(max_tokens, 0, -1):
+        truncated_text_tokens = tokens[:token_count]
+        actual_token_count = token_count_utils.count_token(" ".join(truncated_text_tokens))
```

Comment:
Since we can use tiktoken to actually tokenize the input wouldnt it make sense to use that to truncate rather than tokenizing it over and over and calculating the length?

Reply:
Can we tokenize it, truncate the token sequence, then transform it back to text? If so, that might work... but we'll have to be a little careful about word boundaries (since a token can very-often be sub-word)

Reply:
Oof yeah you're right. That's a real pain

Reply:
I'm wondering if we even need to worry about tokenizing. Models can tokenize slightly differently, so it can vary depending on the model and token_count_utils.count_token is defaulting to gpt3.5. I could simplify this tremendously if we are good with approximating tokens via `text.split()`. The five lines in the for loop becomes one line `truncated_tokens = tokens[:max_tokens]`. I would prefer to simplify it.

Reply:
There is often more than 1 token per word, so if we went with that scheme it would differ a lot. I think that matters especially for handling what we do at that limit. You're right the current method is definitely suboptimal. For now I think what you have is good but I can see us adding functionality to tokenize based on the model being used

Reply:
That makes sense. It would be a valuable improvement.